db/new_seeds/base/add_declarations.rb

```diff
@@ -1,72 +1,201 @@
       induction_record = participant_profile.current_induction_record
+      next unless induction_record # Skip if no induction record
+
       schedule = induction_record.schedule
-      next if Induction::FindBy.call(participant_profile:, lead_provider:, schedule:).blank?
-
-      # Possible declaration states.
-      states = ParticipantDeclaration.states.keys
-      fundable_states = %w[payable paid awaiting_clawback clawed_back]
-      states.reject! { |state| state.in?(fundable_states) } unless participant_profile.fundable?
-      state = states.sample
-
-      # Possible declaration types.
-      declaration_types = %w[started retained-1 retained-2 retained-3 retained-4 extended-1 extended-2 extended-3 completed]
-      existing_declaration_types = participant_profile.participant_declarations.pluck(:declaration_type)
-      declaration_types.reject! { |type| type.in?(existing_declaration_types) }
-      declaration_types.reject! { |type| type.match?(/retained|extended/) } if participant_profile.mentor? && schedule.cohort.mentor_funding?
-      declaration_types.reject! { |type| type.match?(/extended/) } unless schedule.schedule_identifier.match?(/extended/)
-
-      declaration_types.each do |declaration_type|
-        # Keep creating declarations with a 50% chance.
-        break unless Faker::Boolean.boolean(true_ratio: 0.5)
-
-        # Ensure there is a milestone for the declaration type.
-        FactoryBot.create(:milestone, schedule:, declaration_type:, start_date: 1.day.ago) unless schedule.milestones.exists?(declaration_type:)
-
-        # We can't create a declaration if the schedule milestone is in the future,
-        # so we travel to the milestone start_date first.
-        milestone_start_date = schedule.milestones.find_by(declaration_type:).start_date
-
-        travel_to milestone_start_date do
-          # Create declaration of correct type.
-          type = participant_profile.ect? ? :ect : :mentor
-          FactoryBot.create(
-            "#{type}_participant_declaration",
-            state,
+      next unless schedule # Skip if no schedule
+
+      # Check if participant has an induction for this lead provider and schedule
+      next if Induction::FindBy.call(
+        participant_profile:,
+        lead_provider:,
+        schedule:,
+      ).blank?
+
+      # Group by schedule to create milestones efficiently
+      profiles_by_schedule[schedule] ||= []
+      profiles_by_schedule[schedule] << participant_profile
+    end
+
+    # Process each schedule group
+    profiles_by_schedule.each do |schedule, profiles|
+      # Pre-create all needed milestones for this schedule in one batch
+      needed_milestone_types = Set.new
+
+      profiles.each do |participant_profile|
+        profile_declaration_types = declaration_types.dup
+
+        # Filter out existing declaration types
+        existing_types = participant_profile.participant_declarations.pluck(:declaration_type)
```

Comment:
we have a bit of duplication here with similar logic below on line 146 and continuing, can we try and avoid that?

Reply:
Hi @cwrw I've removed the loop as the milestones are being created with the schedules at the top of the script so no need for that inner loop anymore. tks